pkg/apis/crd/v1alpha2/types.go

```diff
@@ -221,3 +226,46 @@ type EgressList struct {
 
 	Items []Egress `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExternalIPPool defines one or multiple IP sets that can be used in external network. For instance, the IPs can be
+// allocated to the Egress resources as the Egress IPs.
+type ExternalIPPool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the ExternalIPPool.
+	Spec ExternalIPPoolSpec `json:"spec"`
+}
+
+type ExternalIPPoolSpec struct {
+	// The IP ranges of this IP pool, e.g. 10.10.0.0/24, 10.10.10.2-10.10.10.20, 10.10.10.30-10.10.10.30.
+	IPRanges []IPRange `json:"ipRanges"`
+	// The Nodes that the external IPs can be assigned to. If empty, it means all Nodes.
```

Comment:
If it is nil, do you think we can say it means auto-assignement is not enabled?

Reply:
I feel it wouldn't be very useful in practice. It seems if users want to manage the egress IPs themselves, providing an ExternalIPPool here and associate it with the Egresses is not as convenient as just specifying the IP in Egress directly. Some benefits when using nil to represent all Nodes in my mind:
1. For a cluster that all nodes are in same subnet, it's more convinient to use nil to select all Nodes as the Egress Nodes. I don't know a typical expression that selects all Nodes without adding labels to them. (unless using `NotIn` and `DoesNotExist` with some meaningless key/value)
2. Have a clear boundary between enabling assignment/failover and disabling it by the externalIPPool field.
3. Keep backwards-compatible with the previous static Egresses.

Reply:
What you said make sense to me. But do you think it is a convention to use empty selector (but not nil) to express selecting all?

Reply:
It seems that it's quite unusual for a nil selector to mean "select all". Is there an example in the K8s API of such behavior? On the other hand, it seems that empty (but not nil) to select all is fairly common. My preference if we do not think there will be other mechanisms / fields for selection would be to make this field *not* optional, with empty meaning "select all" (like the `PodSelector` field in the NetworkPolicy Spec). As Jianjun mentioned in the meeting, if we plan to have other selector fields (potentially mutually-exclusive), that's potentially a different story...

Reply:
I missed the difference between nil and empty, thought they are same somehow. Yes, what you suggested makes sense. Empty is more common to to represent selecting All. I have made the field mandatory in the latest patch. If we have other fields to select Nodes in the future, I think changing a mandatory field to optional is compatible.

Reply:
I think last time we discussed changing from mandatory to optional is not a compatible change. But anyway if we see no use case to make it optional, we can keep the current design (otherwise maybe change back to a pointer, but validating it cannot be nil?).

Reply:
I also believe that changing from mandatory to optional is not considered a compatible changes, since it can break older clients which read `ExternalIPPoolSpec` resources created by a newer client? https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility

Given that this is an alpha API, maybe we should go with what makes the most sense right now, which I would say is making the field mandatory (current proposal).

Reply:
My mistake, was considering the server side storage only. Agree we should keep what makes the most sense right now when we don't have clear idea about how to extend it.